docs/development/search.rst

```diff
@@ -0,0 +1,99 @@
+Search
+============
+
+Read The Docs uses Elasticsearch_ instead of the built in Sphinx search for providing better search
+results. Documents are indexed in the Elasticsearch index and the search is made through the API.
+All the Search Code is open source and lives in the `Github Repository`_.
+Currently we are using the `Elasticsearch 6.3`_ version.
+
+Local Development Configuration
+-------------------------------
+
+Installing and running Elasticsearch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+You need to install and run Elasticsearch_ version 6.3 on your local development machine.
+You can get the installation instructions
+`here <https://www.elastic.co/guide/en/elasticsearch/reference/6.3/install-elasticsearch.html>`_.
+Otherwise, you can also start an Elasticsearch Docker container by running the following command::
+
+    docker run -p 9200:9200 -p 9300:9300 \
+           -e "discovery.type=single-node" \
+           docker.elastic.co/elasticsearch/elasticsearch:6.3.2
+
+Indexing into Elasticsearch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+For using search, you need to index data to the Elasticsearch Index. Run `reindex_elasticsearch`
+management command::
+
+    ./manage.py reindex_elasticsearch
+
+Auto Indexing
+^^^^^^^^^^^^^
+By default, Auto Indexing is turned off in development mode. To turn it on, change the
+`ELASTICSEARCH_DSL_AUTOSYNC` settings to `True` in the `readthedocs/settings/dev.py` file.
+After that, whenever a documentation successfully builds, or project gets added,
+the search index will update automatically.
+
+
+Architecture
+------------
+The search architecture is devided into 2 parts.
+One part is responsible for **indexing** the documents and projects and
+the other part is responsible for querying the Index to show the proper results to users.
+We use the `django-elasticsearch-dsl`_ package mostly to the keep the search working.
+`django-elasticsearch-dsl`_ is a wrapper around `elasticsearch-dsl`_ for easy configuration
+with Django.
+
+Indexing
+^^^^^^^^
+All the Sphinx documents are indexed into Elasticsearch after the build is successful.
+Currently, we do not index MkDocs documents to elasticsearch, but
+`any kind of help is welcome <https://github.com/rtfd/readthedocs.org/issues/1088>`_.
+
+How we index documentations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After any build is successfully finished, `HTMLFile` objects are created for each of the
+`HTML` files and the old version's `HTMLFile` object is deleted. The Signal_
+`bulk_post_create` is dispatched for created and `bulk_post_delete` is dispatched for deleted
+files. Both of the signals are dispatched with the list of the instances of `HTMLFile`
+in `instance_list` parameter.
+
+We listen to the `bulk_post_create` and `bulk_post_delete` signals in our `Search` application and
+index/delete the documentation content from the `HTMLFile` instances.
```

Comment:
Again, it's probably worth mentioning why we designed it this way, because by default it was doing an HTTP request per object.